build_tools/azure/update_environments_and_lock_files.py

```diff
@@ -0,0 +1,416 @@
+            "matplotlib": "min",
+            "threadpoolctl": "2.2.0",
+        },
+    },
+    {
+        "build_name": "py38_conda_forge_openblas_ubuntu_1804",
+        "platform": "linux-64",
+        "channel": "conda-forge",
+        "conda_dependencies": common_dependencies_without_coverage + ["ccache"],
+        "package_constraints": {"python": "3.8", "blas": "[build=openblas]"},
+    },
+    {
+        "build_name": "pylatest_pip_openblas_pandas",
+        "platform": "linux-64",
+        "channel": "defaults",
+        "conda_dependencies": ["python", "ccache"],
+        "pip_dependencies": remove_from(common_dependencies, ["python", "blas"])
+        + docstring_test_dependencies
+        + ["lightgbm", "scikit-image"],
+        "package_constraints": {"python": "3.9"},
+    },
+    {
+        "build_name": "pylatest_pip_scipy_dev",
+        "platform": "linux-64",
+        "channel": "defaults",
+        "conda_dependencies": ["python", "ccache"],
+        "pip_dependencies": remove_from(
+            common_dependencies,
+            [
+                "python",
+                "blas",
+                "matplotlib",
+                "pyamg",
+                # all the dependencies below have a development version
+                # installed in the CI, so they can be removed from the
+                # environment.yml
+                "numpy",
+                "scipy",
+                "pandas",
+                "cython",
+                "joblib",
+                "pillow",
+            ],
+        )
+        + docstring_test_dependencies
+        # python-dateutil is a dependency of pandas and pandas is removed from
+        # the environment.yml. Adding python-dateutil so it is pinned
+        + ["python-dateutil"],
+    },
+    {
+        "build_name": "pypy3",
+        "platform": "linux-64",
+        "channel": "conda-forge",
+        "conda_dependencies": ["pypy"]
+        + remove_from(
+            common_dependencies_without_coverage, ["python", "pandas", "pillow"]
+        )
+        + ["ccache"],
+        "package_constraints": {"blas": "[build=openblas]"},
+    },
+]
+
+
+pip_build_metadata_list = [
+    {
+        "build_name": "debian_atlas_32bit",
+        "pip_dependencies": ["cython", "joblib", "threadpoolctl", "pytest"],
+        "package_constraints": {
+            "joblib": "min",
+            "threadpoolctl": "2.2.0",
+            "pytest": "min",
+            # no pytest-xdist because it causes issue on 32bit
+        },
+        # same Python version as in debian-32 build
+        "python_version": "3.9.2",
+    },
+    {
+        "build_name": "ubuntu_atlas",
+        "pip_dependencies": [
+            "cython",
+            "joblib",
+            "threadpoolctl",
+            "pytest",
+            "pytest-xdist",
+        ],
+        "package_constraints": {"joblib": "min", "threadpoolctl": "min"},
+        # same Python version as in Ubuntu 20.04 build
+        "python_version": "3.8.2",
```

Comment:
With this change, the script works on my m1 machine:

```suggestion
        "python_version": "3.8.13",
```

Reply:
I put 3.8.5 which is available on osx-arm64 on conda-forge. I added a comment saying that this should not matter for pip-compile